File: simple_ros_components/actions/action_client_py/action_client_py/add_until_client.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from rclpy.action import ActionClient
from action_interfaces.action import AddUntil
from rclpy.action.client import ClientGoalHandle, GoalStatus


class AddUntilClientNode(Node):
    def __init__(self):
        super().__init__("add_until_client_node")
        self.add_until_client_ = ActionClient(
            node=self, action_type=AddUntil, action_name="AddUntil"
        )

        # Declare ROS parameters with default values
        self.declare_parameter("target_number", 10)
        self.declare_parameter("period", 0.5)

    def send_goal(self, target_number, period):
        """
        Send a goal to the AddUntil action server.

        Args:
            target_number (int): The number up to which values will be added.
            period (float): Time delay (in seconds) between additions.
        """

        # wait for server to be available
        is_server_ready_ = self.add_until_client_.wait_for_server(5)

        if not is_server_ready_:
            self.get_logger().error("Server is not ready...")
            return

        self.get_logger().info("Server is ready...")

        # Populate the goal message
        goal = AddUntil.Goal()
        goal.target_number = target_number
        goal.period = period

        self.get_logger().info("Sending command to server...")

        # Send the goal asynchronously
        response_handle = self.add_until_client_.send_goal_async(
            goal=goal, feedback_callback=self.goal_feedback_callback
        )

        # Register callback for goal acceptance/rejection
        response_handle.add_done_callback(self.goal_response_callback)

    def goal_feedback_callback(self, feedback):
        """
        Callback function for receiving feedback from the action server.

        Args:
            feedback (AddUntil.FeedbackMessage): Feedback message containing
            the current intermediate sum.
        """
        self.get_logger().info(f"Feedback: {feedback.feedback.intermediate_sum}")

        # For testing purpose, triger cancel request
        # if feedback.feedback.intermediate_sum == 91:
        #     print("Canceling goal")
        #     self.goal_handle_.cancel_goal_async()

    def goal_response_callback(self, future):
        """
        Callback function for handling the goal response.

        This is called once the server accepts or rejects the goal.

        Args:
            future: Future object containing the ClientGoalHandle.
        """
        try:
            self.goal_handle_: ClientGoalHandle = future.result()
        except Exception as e:
            self.get_logger().error(f"Goal request exception: {e}")
            return

        if self.goal_handle_ is None:
            self.get_logger().error("Invalid goal handle")
            return

        if not self.goal_handle_.accepted:
            self.get_logger().warning("Goal got rejected")
            return

        self.get_logger().info("Goal accepted")
        # Request the final result asynchronously
        self.goal_handle_.get_result_async().add_done_callback(
            self.goal_result_callback
        )

    def goal_result_callback(self, future):
        """
        Callback function for handling the final result of the action.

        Args:
            future: Future object containing the result and status.
        """

        # check status of request
        status = future.result().status
        result = future.result().result

        if status == GoalStatus.STATUS_SUCCEEDED:
            self.get_logger().info("Succeeded goal")
        elif status == GoalStatus.STATUS_ABORTED:
            self.get_logger().error("Aborted goal")
        elif status == GoalStatus.STATUS_CANCELED:
            self.get_logger().info("Canceled goal")

        self.get_logger().info(f"Client result = {result.sum}")


def main(args=None):
    rclpy.init(args=args)

    node = AddUntilClientNode()

    target_number = node.get_parameter("target_number").value
    period = node.get_parameter("period").value

    node.send_goal(target_number, period)

    rclpy.spin(node)

    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: simple_ros_components/actions/action_client_cpp/src/add_until_client.cpp
#include "action_client_cpp/add_until_client.hpp"

AddUntilClient::AddUntilClient() : Node("add_until_client") {
  add_until_client_ = rclcpp_action::create_client<AddUntil>(this, "AddUntil");
}

void AddUntilClient::send_goal(
    AddUntil::Goal::_target_number_type target_number,
    AddUntil::Goal::_period_type period) {
  // wait for server to be available. if the server is not available after 5
  // seconds, shutdown the client.
  if (!add_until_client_->wait_for_action_server(std::chrono::seconds(5))) {
    RCLCPP_ERROR(this->get_logger(),
                 "Action server not available after waiting for 5 seconds");
    rclcpp::shutdown();
    return;
  }

  // create goal
  auto goal_msg = AddUntil::Goal();
  goal_msg.target_number = target_number;
  goal_msg.period = period;

  auto send_goal_options = rclcpp_action::Client<AddUntil>::SendGoalOptions{};

  // define result callback
  // can also use std::bind() instead of lambda
  send_goal_options.result_callback =
      [this](const rclcpp_action::ClientGoalHandle<AddUntil>::WrappedResult
                 &result) {
        auto status = result.code;
        if (status == rclcpp_action::ResultCode::SUCCEEDED) {
          RCLCPP_INFO(this->get_logger(), "Goal succeeded");
        } else if (status == rclcpp_action::ResultCode::ABORTED) {
          RCLCPP_WARN(this->get_logger(), "Goal was aborted");
        } else if (status == rclcpp_action::ResultCode::CANCELED) {
          RCLCPP_WARN(this->get_logger(), "Goal was canceled");
        } else {
          RCLCPP_ERROR(this->get_logger(), "Unknown result code");
        }

        RCLCPP_INFO(this->get_logger(), "Result received: %ld",
                    result.result->sum);

        // shutdown client after receiving result
        rclcpp::shutdown();
      };

  send_goal_options.goal_response_callback =
      [this](const rclcpp_action::ClientGoalHandle<AddUntil>::SharedPtr
                 &goal_handle) {
        if (!goal_handle) {
          RCLCPP_ERROR(this->get_logger(), "Goal was rejected by server");

          // shutdown client since the goal was rejected.
          rclcpp::shutdown();
          return;
        } else {
          RCLCPP_INFO(this->get_logger(),
                      "Goal accepted by server, waiting for result");
        }
      };

  send_goal_options.feedback_callback =
      [this](const rclcpp_action::ClientGoalHandle<AddUntil>::SharedPtr
                 &goal_handle,
             const std::shared_ptr<const AddUntil::Feedback> feedback) {
        (void)goal_handle;
        RCLCPP_INFO(this->get_logger(), "Received feedback: %ld",
                    feedback->intermediate_sum);

        // Simulate canceling the goal based on feedback
        // if (feedback->intermediate_sum > 20) {
        //   RCLCPP_INFO(this->get_logger(),
        //               "Canceling goal because intermediate sum is too high");
        //   add_until_client_->async_cancel_goal(goal_handle);
        // }
      };

  // send goal
  RCLCPP_INFO(this->get_logger(), "Sending goal");
  add_until_client_->async_send_goal(goal_msg, send_goal_options);
}
